### PR TITLE
[Feature] IAM supports identity acl for console and api access

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
- API are provided for the administrator to modify and list the ACL for console access and api access.
- For acl create or update for console access and api access:
  - The netmask should be a valid CIDR and the ip range cannot be set repeatedly.
  - Up to 200 CIDR and ip range can be created in total.
  - Default ip range is 0.0.0.0-255.255.255.255.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
```
NONE
```

**Special notes for your reviewer**:

**Release note**:

```release-note
1. New identity acl resource, test and document.
2. Add ip range check into validate functions.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed
Identity ACL test for Console access
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccIdentitACL_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccIdentitACL_basic -timeout 360m -parallel 4
=== RUN   TestAccIdentitACL_basic
=== PAUSE TestAccIdentitACL_basic
=== CONT  TestAccIdentitACL_basic
--- PASS: TestAccIdentitACL_basic (41.21s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       41.357s
```
Identity ACL test for API access
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccIdentitACL_apiAccess'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccIdentitACL_apiAccess -timeout 360m -parallel 4
=== RUN   TestAccIdentitACL_apiAccess
=== PAUSE TestAccIdentitACL_apiAccess
=== CONT  TestAccIdentitACL_apiAccess
--- PASS: TestAccIdentitACL_apiAccess (36.92s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       36.999s
```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -129,12 +129,12 @@ github.com/hashicorp/yamux v0.0.0-20180604194846-3520598351bb h1:b5rjCoWHc7eqmAS
...
-github.com/huaweicloud/golangsdk v0.0.0-20210310091354-6e2f74a11d0f h1:AFZM6Vdn8dS764DaaiRKGP26iM7ta2zlQqM+SR1svGk=
-github.com/huaweicloud/golangsdk v0.0.0-20210310091354-6e2f74a11d0f/go.mod h1:fcOI5u+0f62JtJd7zkCch/Z57BNC6bhqb32TKuiF4r0=
-github.com/jen20/awspolicyequivalence v1.1.0 h1:cn37D6o0lXLwqx2neCokGfaB3LLNSo5CrLMLGjY609g=
-github.com/jen20/awspolicyequivalence v1.1.0/go.mod h1:PV1fS2xyHhCLp83vbgSMFr2drM4GzG61wkz+k4pOG3E=
+github.com/huaweicloud/golangsdk v0.0.0-20210227060920-53016d30a39b h1:nrUhoqO5BbrPxuWUIVj5Wt/oJoBzyzLiBHVGCCF33tg=
+github.com/huaweicloud/golangsdk v0.0.0-20210227060920-53016d30a39b/go.mod h1:fcOI5u+0f62JtJd7zkCch/Z57BNC6bhqb32TKuiF4r0=
+github.com/huaweicloud/golangsdk v0.0.0-20210311035428-8cfccba69a01 h1:NIh2E9JUKO2/4xGjcSFgzXLp/cCcyPj6mVmPZlMzo/Q=
+github.com/huaweicloud/golangsdk v0.0.0-20210311035428-8cfccba69a01/go.mod h1:fcOI5u+0f62JtJd7zkCch/Z57BNC6bhqb32TKuiF4r0=
+github.com/jen20/awspolicyequivalence v0.0.0-20170831201602-3d48364a137a h1:FyS/ubzBR5xJlnJGRTwe7GUHpJOR4ukYK3y+LFNffuA=
+github.com/jen20/awspolicyequivalence v0.0.0-20170831201602-3d48364a137a/go.mod h1:uoIMjNxUfXi48Ci40IXkPRbghZ1vbti6v9LCbNqRgHY=
```
Comment (ShiChangkuo) [inline go.sum]: please rebase the latest `master` branch.

File: huaweicloud/provider_test.go [hunk 2]
```diff
@@ -61,6 +61,12 @@ func testAccPreCheck(t *testing.T) {
 	}
 }
 
+func testAccPrechecRegion(t *testing.T) {
+	if HW_REGION_NAME == "" {
```
Comment (ShiChangkuo) [inline huaweicloud/provider_test.go]: `HW_REGION_NAME` is required for all acctests, so I think this check is unnecessary

File: huaweicloud/resource_huaweicloud_identity_acl.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceIdentityACLCreate(d *schema.ResourceData, meta interface{}) error {
+	updateACLPolicy(d, meta)
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl.go]: here should check the result.

File: huaweicloud/resource_huaweicloud_identity_acl.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	return resourceIdentityACLRead(d, meta)
+}
+
+func resourceIdentityACLRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	iamClient, err := config.IAMV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud iam client: %s", err)
+	}
+	domainID := config.DomainID
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl.go]: using d.Id() will be better

File: huaweicloud/resource_huaweicloud_identity_acl.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+		}
+		d.Set("ip_ranges", ipRanges)
+	}
+
+	return nil
+}
+
+func resourceIdentityACLUpdate(d *schema.ResourceData, meta interface{}) error {
+	if d.HasChanges("ip_cidrs", "ip_ranges") {
+		updateACLPolicy(d, meta)
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl.go]: here should check the result.

File: huaweicloud/resource_huaweicloud_identity_acl.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	return resourceIdentityACLRead(d, meta)
+}
+
+func resourceIdentityACLDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	iamClient, err := config.IAMV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud iam client: %s", err)
+	}
+	domainID := config.DomainID
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl.go]: using d.Id() will be better

File: huaweicloud/resource_huaweicloud_identity_acl_test.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+)
+
+func TestAccIdentitACL_basic(t *testing.T) {
+	var acl acl.ACLPolicy
+	resourceName := "huaweicloud_identity_acl.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccPrechecRegion(t)
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl_test.go]: as iam should have the admin privileges, please use `testAccPreCheckAdminOnly`.

File: huaweicloud/resource_huaweicloud_identity_acl_test.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+func TestAccIdentitACL_apiAccess(t *testing.T) {
+	var acl acl.ACLPolicy
+	resourceName := "huaweicloud_identity_acl.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccPrechecRegion(t)
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl_test.go]: ditto.

File: huaweicloud/validators.go [hunk 9]
```diff
@@ -95,6 +95,42 @@ func validateCIDR(v interface{}, k string) (ws []string, errors []error) {
...
+		} else {
+			netIPs = append(netIPs, ipnet)
+		}
+	}
+	if len(errors) == 0 {
+		if netIPs[0].Equal(netIPs[1]) {
+			errors = append(errors, fmt.Errorf("Two network IP address of %q cannot equal, got %q", k, value))
+		}
+		ipIndexArr := []int{12, 13, 14, 15}
+		for _, index := range ipIndexArr {
```
Comment (ShiChangkuo) [inline huaweicloud/validators.go]: s/ipIndexArr/len(netIPs[0])/

File: huaweicloud/provider_test.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/provider_test.go]: Thanks, testAccPrecheckRegion is removed from provider_test.

File: huaweicloud/validators.go (hunk 9 above)
Comment (Lance52259) [reply huaweicloud/validators.go]: Use a for loop to compare the four parts of IPv4 address in turn now.
It's easy to read than range 12 to 15 of net.IP.
Split string to arrays have less time consuming.

File: huaweicloud/resource_huaweicloud_identity_acl_test.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl_test.go]: Added `testAccPreCheckAdminOnly`.

File: huaweicloud/resource_huaweicloud_identity_acl_test.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl_test.go]: Added `testAccPreCheckAdminOnly`.

File: huaweicloud/resource_huaweicloud_identity_acl.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl.go]: Thanks, now using d.Id() to get ACL id and assigned to domain Id.

File: go.sum (hunk 1 above)
Comment (Lance52259) [reply go.sum]: Done

File: huaweicloud/resource_huaweicloud_identity_acl.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl.go]: thanks, I forgot to check the error, that was a mistake.

File: huaweicloud/resource_huaweicloud_identity_acl.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl.go]: sure, it is unnecessary to get the domain id twice.

File: huaweicloud/resource_huaweicloud_identity_acl.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl.go]: done

File: huaweicloud/resource_huaweicloud_identity_acl.go [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+	return resourceIdentityACLRead(d, meta)
+}
+
+func resourceIdentityACLRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	iamClient, err := config.IAMV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud iam client: %s", err)
+	}
+	domainID := d.Id()
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_identity_acl.go]: d.Id() will not be empty in Read function.

File: docs/resources/identity_acl.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    description = "This is a test ip range"
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `acl_type` - (Required, String, ForceNew) Specifies the ACL is created through the console or API.
```
Comment (ShiChangkuo) [inline docs/resources/identity_acl.md]: suggest to rename it to `type`

File: huaweicloud/resource_huaweicloud_identity_acl.go (hunk 10 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_identity_acl.go]: IAM resources cannot be configured without domain id. It has pre-check for domain id in LoadAndValidate function.
So, I think domain id will not be empty in create function too.

File: docs/resources/identity_acl.md (hunk 11 above)
Comment (Lance52259) [reply docs/resources/identity_acl.md]: yeah, `type` is better than `acl_type`. Do you think `ip_cidrs` need to rename?

**Pull request merged**